### ISSUE TITLE
Roslyn doesn't update value set to properties with outputs stricter that inputs if it's done in object initializer

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#41002]
**Version Used**: 
```
commit 1d4cf98ece59ab4365adcc7d414d3c758010d698
Merge: 1302d42173 26187bd0d8
Author: Jason Malinowski <jason.malinowski@microsoft.com>
Date:   Mon Jan 13 12:18:17 2020 -0800

    Merge pull request #40897 from jasonmalinowski/null-annotate-quick-info

    Null annotate Quick Info
```

**Steps to Reproduce**:

Compile the following code:
```cs
using System.Diagnostics.CodeAnalysis;
public class C
{
    [AllowNull] public C P { get; set; } = null!;
}

class Program
{
    static void M()
    {
        var c1 = new C();
        c1.P = null;
        c1.P.ToString(); // no warnings

        var c2 = new C
        {
            P = null
        };
        c2.P.ToString(); // warning
    }
}
```

**Expected Behavior**:
No warnings. The two code samples are equivalent.

**Actual Behavior**:
`warning CS8602: Dereference of a possibly null reference.` is reported for `c2.P.ToString` but not for `c1.P.ToString`
